sp/src/public/vscript/ivscript.h

```diff
@@ -927,14 +961,13 @@ class IScriptVM
 	//--------------------------------------------------------
 
 #ifdef MAPBASE_VSCRIPT
-	// When a RegisterInstance instance is deleted, VScript normally treats it as a strong reference and only deregisters the instance itself, preserving the registered data
-	// it points to so the game can continue to use it.
-	// bAllowDestruct is supposed to allow VScript to treat it as a weak reference created by the script, destructing the registered data automatically like any other type.
-	// This is useful for classes pretending to be primitive types.
-	virtual HSCRIPT RegisterInstance( ScriptClassDesc_t *pDesc, void *pInstance, bool bAllowDestruct = false ) = 0;
+	// if bRefCounted is true, pInstance memory will be deleted by the script,
+	// returning the result will then behave as if the instance was constructed in script.
+	// Functions that return the result of this need to return HSCRIPT_RC
+	virtual HSCRIPT RegisterInstance( ScriptClassDesc_t *pDesc, void *pInstance, bool bRefCounted = false ) = 0;
```

Comment:
Adding `RegisterInstanceRefcounted` would mean also duplicating the `SquirrelVM::RegiserInstance` function. A template or a wrapper could work for simplification.

I like the non-explicit nature of `HSCRIPT_RC`. It's still a valid `HSCRIPT` object; it can be called, cached, deleted. It doesn't require any change other than return type. I also thought of specifying the refcounted nature in function registration, but that was too noisy.

Reply:
As it stands, the programmer must take care to convert the return-value to `HSCRIPT_RC` - as the comment mentions. This could be enforced by the type-system instead.
Anyway, this is just a drive-by review. I didn't spot a problematic case in the changes FWIW. Feel free to mark as resolved.

Reply:
I think I understand what you mean. I added `HSCRIPT_RC RegisterInstanceRefCounted()`, removed `bRefCounted` parameter and `HSCRIPT_RC::operator HSCRIPT()`, added `ScriptVariant_t::operator=( HSCRIPT_RC )`. And it works to enforce return type, but it breaks passing `HSCRIPT_RC` to `HSCRIPT` parameters to call, copy or delete. I'll just leave it as is without a clean solution for now.